app/src/main/java/bunkou/tms/HomeWork11/WidgetWeather.kt

```diff
@@ -0,0 +1,116 @@
+package bunkou.tms.HomeWork11
+
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.view.View
+import android.widget.RemoteViews
+import bunkou.tms.HomeWork11.entity.CurrentWeather
+import bunkou.tms.HomeWork11.mappers.WeatherMapper
+import bunkou.tms.HomeWork11.retrofit.RetrofitWeatherFactory
+import bunkou.tms.R
+import com.squareup.picasso.Picasso
+import kotlinx.coroutines.*
+import kotlin.math.roundToInt
+
+private const val WIDGET_REQUEST_CODE = 2121
+
+class WeatherAppWidget : AppWidgetProvider() {
+
+    override fun onUpdate(
+        context: Context,
+        appWidgetManager: AppWidgetManager,
+        appWidgetIds: IntArray
+    ) {
+        for (appWidgetId in appWidgetIds) {
+            updateAppWidget(context, appWidgetManager, appWidgetId)
+        }
+    }
+
+    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
+        // When the user deletes the widget, delete the preference associated with it.
+        for (appWidgetId in appWidgetIds) {
+            deleteTitlePref(context, appWidgetId)
+        }
+    }
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        super.onReceive(context, intent)
+        val appWidgetManager = AppWidgetManager.getInstance(context)
+        if (context != null) {
+            val appWidgetIds = appWidgetManager.getAppWidgetIds(
+                ComponentName(
+                    context,
+                    WeatherAppWidget::class.java
+                )
+            )
+            if (appWidgetIds != null) {
+                for (appWidgetId in appWidgetIds) {
+                    updateAppWidget(context, appWidgetManager, appWidgetId)
+                }
+            }
+        }
+    }
+}
+
+internal fun updateAppWidget(
+    context: Context,
+    appWidgetManager: AppWidgetManager,
+    appWidgetId: Int
+) {
+    // Construct the RemoteViews object
+    val remoteViews = RemoteViews(context.packageName, R.layout.weather_app_widget)
+    addWidgetClickListener(context, remoteViews)
+    fun launchIo(task: suspend () -> Unit) {
+        CoroutineScope(Dispatchers.IO).launch {
+            task()
+        }
+    }
```

Comment:
I think it is better to put this function in a separate file

Reply:
Extract to Utils.kt